[SQL] RustError: `is_distinct__` expects `T` but gets `Option<T>`

SQL:

```sql
CREATE TABLE t1(c0 varchar) with ('materialized' = 'true');
CREATE TABLE t2(c0 varchar) with ('materialized' = 'true');

CREATE MATERIALIZED VIEW v2 AS (SELECT 1 FROM t1 LEFT JOIN t2 ON ((CAST('8288-9-27 0:16:31' AS TIMESTAMP))IS NOT DISTINCT FROM(NULL)) WHERE true);
```
The SQL compilation succeeds but Rust fails with the following error:

```rust
Rust error: the Rust code generated based on the SQL (possibly combined with any user-provided UDF Rust and TOML code) failed to compile.
This should not happen (except if the error is in user-provided UDF code).
Please file a bug report with the example SQL and any UDF Rust/TOML that triggers it at:
https://github.com/feldera/feldera/issues


The compilation task exited with status code 101 and produced the following logs:

stderr:
   Compiling project-1b79b744af531660de76519c4ef6a332c3f9a296d8c5fc3f906ee7f7cce69da8 v0.1.0 (/home/abhizer/.feldera/compiler/rust-compilation/projects/project-1b79b744af531660de76519c4ef6a332c3f9a296d8c5fc3f906ee7f7cce69da8)
error[E0308]: mismatched types
    --> projects/project-1b79b744af531660de76519c4ef6a332c3f9a296d8c5fc3f906ee7f7cce69da8/src/main.rs:153:86
     |
153  |             (!(is_distinct__(cast_to_Timestamp_s(String::from("8288-9-27 0:16:31")), None::<Timestamp>).unwrap()))
     |                ------------- arguments to this function are incorrect                ^^^^^^^^^^^^^^^^^ expected `Timestamp`, found `Option<Timestamp>`
     |
     = note: expected struct `feldera_sqllib::Timestamp`
                  found enum `std::option::Option<feldera_sqllib::Timestamp>`
note: function defined here
    --> /mnt/hdd0/Development/feldera.git/main/crates/sqllib/src/lib.rs:1177:8
     |
1177 | pub fn is_distinct__<T>(left: T, right: T) -> bool
     |        ^^^^^^^^^^^^^

error[E0599]: no method named `unwrap` found for type `bool` in the current scope
   --> projects/project-1b79b744af531660de76519c4ef6a332c3f9a296d8c5fc3f906ee7f7cce69da8/src/main.rs:153:105
    |
153 |             (!(is_distinct__(cast_to_Timestamp_s(String::from("8288-9-27 0:16:31")), None::<Timestamp>).unwrap()))
    |                                                                                                         ^^^^^^ method not found in `bool`

Some errors have detailed explanations: E0308, E0599.
For more information about an error, try `rustc --explain E0308`.
error: could not compile `project-1b79b744af531660de76519c4ef6a332c3f9a296d8c5fc3f906ee7f7cce69da8` (bin "project-1b79b744af531660de76519c4ef6a332c3f9a296d8c5fc3f906ee7f7cce69da8") due to 2 previous errors


stdout:
```

--- a/crates/sqllib/src/lib.rs
+++ b/crates/sqllib/src/lib.rs
@@ -1187,11 +1187,7 @@ pub fn is_distinct_N_N<T>(left: Option<T>, right: Option<T>) -> bool
 where
     T: Eq,
 {
-    match (left, right) {
-        (Some(a), Some(b)) => a != b,
-        (None, None) => false,
-        _ => true,
-    }
+    left != right
 }
 
 #[doc(hidden)]

--- a/crates/sqllib/src/operators.rs
+++ b/crates/sqllib/src/operators.rs
@@ -20,16 +20,6 @@ where
 for_all_compare!(eq, bool, T where Eq);
 
 #[doc(hidden)]
-pub fn is_same__<T>(left: T, right: T) -> bool
-where
-    T: Eq,
-{
-    left == right
-}
-
-// There is only one is_same__ function; no macros are needed.
-
-#[doc(hidden)]
 #[inline(always)]
 pub(crate) fn neq<T>(left: T, right: T) -> bool
 where

--- a/sql-to-dbsp-compiler/SQL-compiler/src/main/java/org/dbsp/sqlCompiler/compiler/backend/rust/RustSqlRuntimeLibrary.java
+++ b/sql-to-dbsp-compiler/SQL-compiler/src/main/java/org/dbsp/sqlCompiler/compiler/backend/rust/RustSqlRuntimeLibrary.java
@@ -56,7 +56,6 @@ public class RustSqlRuntimeLibrary {
         this.universalFunctions.put("gt", DBSPOpcode.GT);
         this.universalFunctions.put("lte", DBSPOpcode.LTE);
         this.universalFunctions.put("gte", DBSPOpcode.GTE);
-        this.universalFunctions.put("is_same", DBSPOpcode.IS_NOT_DISTINCT);
         this.universalFunctions.put(DBSPOpcode.IS_DISTINCT.toString(), DBSPOpcode.IS_DISTINCT);
         this.universalFunctions.put(DBSPOpcode.MIN.toString(), DBSPOpcode.MIN);
         this.universalFunctions.put(DBSPOpcode.MAX.toString(), DBSPOpcode.MAX);
@@ -163,7 +162,7 @@ public class RustSqlRuntimeLibrary {
             opcode == DBSPOpcode.MAX || opcode == DBSPOpcode.MIN ||
             opcode == DBSPOpcode.AGG_GTE || opcode == DBSPOpcode.AGG_LTE ||
             opcode == DBSPOpcode.AGG_MIN || opcode == DBSPOpcode.AGG_MAX ||
-            opcode == DBSPOpcode.IS_DISTINCT || opcode == DBSPOpcode.IS_NOT_DISTINCT) {
+            opcode == DBSPOpcode.IS_DISTINCT) {
             map = this.universalFunctions;
         } else if (ltype.as(DBSPTypeBool.class) != null) {
             map = this.booleanFunctions;
@@ -203,11 +202,6 @@ public class RustSqlRuntimeLibrary {
             suffixl = ltype.nullableSuffix();
             assert rtype != null;
             suffixr = rtype.nullableSuffix();
-            if (opcode == DBSPOpcode.IS_NOT_DISTINCT || opcode == DBSPOpcode.IS_DISTINCT) {
-                // Argument types always match
-                suffixl = "";
-                suffixr = "";
-            }
         } else if (opcode == DBSPOpcode.CONTROLLED_FILTER_GTE) {
             tsuffixl = "";
             tsuffixr = "";

--- a/sql-to-dbsp-compiler/SQL-compiler/src/main/java/org/dbsp/sqlCompiler/compiler/backend/rust/ToRustVisitor.java
+++ b/sql-to-dbsp-compiler/SQL-compiler/src/main/java/org/dbsp/sqlCompiler/compiler/backend/rust/ToRustVisitor.java
@@ -1148,7 +1148,7 @@ public class ToRustVisitor extends CircuitVisitor {
         while (comparator.is(DBSPFieldComparatorExpression.class)) {
             DBSPFieldComparatorExpression fc = comparator.to(DBSPFieldComparatorExpression.class);
             DBSPExpression eq = new DBSPBinaryExpression(node, new DBSPTypeBool(node, false),
-                    DBSPOpcode.IS_NOT_DISTINCT, left.deref().field(fc.fieldNo), right.deref().field(fc.fieldNo));
+                    DBSPOpcode.IS_DISTINCT, left.deref().field(fc.fieldNo), right.deref().field(fc.fieldNo)).not();
             result = new DBSPBinaryExpression(node, eq.getType(), DBSPOpcode.AND, result, eq);
             comparator = fc.source;
         }

--- a/sql-to-dbsp-compiler/SQL-compiler/src/main/java/org/dbsp/sqlCompiler/compiler/frontend/ExpressionCompiler.java
+++ b/sql-to-dbsp-compiler/SQL-compiler/src/main/java/org/dbsp/sqlCompiler/compiler/frontend/ExpressionCompiler.java
@@ -456,7 +456,7 @@ public class ExpressionCompiler extends RexVisitorImpl<DBSPExpression>
             }
             if (opcode.isComparison()) {
                 expressionResultType = DBSPTypeBool.create(anyNull);
-            } else if (opcode == DBSPOpcode.IS_DISTINCT || opcode == DBSPOpcode.IS_NOT_DISTINCT)
+            } else if (opcode == DBSPOpcode.IS_DISTINCT)
                 // Never null
                 expressionResultType = DBSPTypeBool.create(false);
             if (commonBase.is(DBSPTypeNull.class)) {

--- a/sql-to-dbsp-compiler/SQL-compiler/src/main/java/org/dbsp/sqlCompiler/compiler/visitors/inner/Simplify.java
+++ b/sql-to-dbsp-compiler/SQL-compiler/src/main/java/org/dbsp/sqlCompiler/compiler/visitors/inner/Simplify.java
@@ -26,6 +26,7 @@ package org.dbsp.sqlCompiler.compiler.visitors.inner;
 import org.apache.calcite.util.DateString;
 import org.apache.calcite.util.TimeString;
 import org.apache.commons.lang3.StringUtils;
+import org.dbsp.sqlCompiler.circuit.operator.DBSPUnaryOperator;
 import org.dbsp.sqlCompiler.compiler.DBSPCompiler;
 import org.dbsp.sqlCompiler.compiler.visitors.VisitDecision;
 import org.dbsp.sqlCompiler.ir.IDBSPInnerNode;
@@ -425,7 +426,7 @@ public class Simplify extends InnerRewriteVisitor {
                     try {
                         result = lit.to(IsNumericLiteral.class).negate().to(DBSPExpression.class);
                     } catch (ArithmeticException ex) {
-                        // ignore
+                        // leave unchanged
                     }
                 }
             }
@@ -445,6 +446,12 @@ public class Simplify extends InnerRewriteVisitor {
                     result = b;
                 else
                     result = new DBSPBoolLiteral(expression.getNode(), expression.getType(), !b.value);
+            } else if (source.is(DBSPUnaryOperator.class)) {
+                // !!e = e, true in ternary logic too
+                DBSPUnaryExpression unarySource = source.to(DBSPUnaryExpression.class);
+                if (unarySource.operation == DBSPOpcode.NOT) {
+                    result = unarySource.source;
+                }
             }
         }
         this.map(expression, result.cast(expression.getType()));

--- a/sql-to-dbsp-compiler/SQL-compiler/src/main/java/org/dbsp/sqlCompiler/ir/expression/DBSPOpcode.java
+++ b/sql-to-dbsp-compiler/SQL-compiler/src/main/java/org/dbsp/sqlCompiler/ir/expression/DBSPOpcode.java
@@ -40,7 +40,6 @@ public enum DBSPOpcode {
     MIN("min", false),
     CONCAT("||", false),
     IS_DISTINCT("is_distinct", false),
-    IS_NOT_DISTINCT("is_not_distinct", false),
     SQL_INDEX("[]", false),
     MAP_INDEX("[]", false),
     // map index in a variant value

--- a/sql-to-dbsp-compiler/SQL-compiler/src/test/java/org/dbsp/sqlCompiler/compiler/sql/simple/RegressionTests.java
+++ b/sql-to-dbsp-compiler/SQL-compiler/src/test/java/org/dbsp/sqlCompiler/compiler/sql/simple/RegressionTests.java
@@ -1553,4 +1553,12 @@ public class RegressionTests extends SqlIoTest {
                  -------------------""");
         this.addRustTestCase(ccs);
     }
+
+    @Test
+    public void issue3278() {
+        this.compileRustTestCase("""
+                CREATE TABLE t1(c0 varchar);
+                CREATE TABLE t2(c0 varchar);
+                CREATE VIEW v2 AS (SELECT 1 FROM t1 JOIN t2 ON (1 IS NOT DISTINCT FROM NULL));""");
+    }
 }

